src/enum.spec.ts

```diff
@@ -0,0 +1,179 @@
+import { Enum } from "./enum.js";
+import type { Equal, Expect } from "./shared/tests.js";
+
+describe("Enum", () => {
+	test("Default", () => {
+		type Event = Enum<{
+			Open: true;
+			Data: { value: unknown };
+			Close: true;
+		}>;
+
+		const Event = Enum({} as Event);
+
+		{
+			const eventOpen = Event.Open();
+			expect(eventOpen).toStrictEqual({ _type: "Open" });
+			const eventData = Event.Data({ value: "..." });
+			expect(eventData).toStrictEqual({ _type: "Data", value: "..." });
+			const eventClose = Event.Close();
+			expect(eventClose).toStrictEqual({ _type: "Close" });
+
+			void [eventOpen, eventData, eventClose];
+		}
+	});
+
+	test("Mapped", () => {
+		type Event = Enum<{
+			Open: true;
+			Data: { value: unknown };
+			Close: true;
+		}>;
+
+		const Event = Enum({} as Event, {
+			Data: (value: unknown) => ({ value }),
+		});
+
+		{
+			const eventOpen = Event.Open();
+			expect(eventOpen).toStrictEqual({ _type: "Open" });
+			const eventData = Event.Data("...");
+			expect(eventData).toStrictEqual({ _type: "Data", value: "..." });
+			const eventClose = Event.Close();
+			expect(eventClose).toStrictEqual({ _type: "Close" });
+
+			void [eventOpen, eventData, eventClose];
+		}
+	});
+
+	test("Custom", () => {
+		type Event = Enum<
+			{
+				Open: true;
+				Data: { value: unknown };
+				Close: true;
+			},
+			"custom"
+		>;
+
+		const Event = Enum(
+			{} as Event,
+			{
+				Data: (value: unknown) => ({ value }),
+			},
```

Comment:
Need to rework this so that you can set a custom discriminant without needing a placeholder/empty mapper positional argument.

Perhaps I really do need to explore the `Enum.on(discriminant)` interface to allow for a unified interface for all runtime Enum utils:

```ts
// default
type Foo = Enum<{ ... }>
const Foo = Enum({} as Foo)
const Foo = Enum({} as Foo, { ... })
if (Enum.is(value, "...")) { ... }
const result = Enum.match(value, { ... })

// custom
type Foo = Enum<{ ... }, "custom">
const Foo = Enum.on("custom")({} as Foo)
const Foo = Enum.on("custom")({} as Foo, { ... })
if (Enum.on("custom").is(value, "...")) { ... }
const result = Enum.on("custom").match(value, { ... })
```

And I can't use `discriminant` as optional first argument because that breaks `value is ...` type guard functionality for `Enum.is(...)`.

And I can't use a tuple `[value, discriminant]` as `target` argument because that also breaks the `Enum.is(...)` `value is ...` type guard functionality.

And using an `options` object argument would obscure `if` statements and `match` expressions too match.

Reply:
And I think the order of the arguments supporting custom discriminants should be `value` then `discriminant` so that the `discriminant` parameter can have auto-complete based on the available fields of `value`, rather than starting as `string` and then narrowing based on a second position `value` parameter.

Also, I want to avoid runtime generation of intermediate closure and functions, like with the `.on` approach.

Would be ideal to have everything as static as possible, e.g.:

```ts
Enum(value, mapper?)
Enum.custom(value, discriminant, mapper?)
Enum.is(value, matcher)
Enum.custom.is(value, discriminant, matcher)
Enum.match(value, matcher)
Enum.custom.match(value, discriminant, matcher)
```

Reply:
> Need to rework this so that you can set a custom discriminant without needing a placeholder/empty mapper positional argument.

But if this is the original problem, then what if I could keep the existing API of everything else and instead make the one-time fire `Enum` runtime constructor accept a chain, or otherwise avoid an optional property?

```ts
Enum(value, discriminant?)
Enum.map(Enum(value, discriminant?), mapper)
```

Although this does seem quite cumbersome and I don't really like it that much.

---

After some playing around, I think I found what feels best:

```ts
Enum.builder(value, mapper?)
Enum.builder_(valuer, discriminant, mapper?)
Enum.is(value, matcher)
Enum.is_(value, discriminant, matcher)
Enum.match(value, matcher)
Enum.match_(value, matcher)
```

Removing the overloading on `Enum()` allows for cleaner discovery of the functions without all the function-related `bind` and `toString` floating in there too. It also allows me to play with re-exporting the functions as an aggregate rather than an object.

---

I've properly dismissed the idea of adding another full namespace for the runtime and maybe the types space too. 

- This is FAR more work because a lot of the utility types will need to be rewritten and their signatures/parameters duplicated between the `Enum$` base version and the `Enum` default version.
- And I don't want to widen the root-level API surface beyond `Enum` and the built-ins.

```ts
import { Enum } from "unenum"

type Foo = Enum<{ ... }>
const Foo = Enum({} as Foo)
Enum.is(value, "...")
Enum.match(value, { ... })

import { Enum$ } from "unenum"

type Foo = Enum$<{ ... }, "custom">
const Foo = Enum$({} as Foo, , "custom", mapper?)
Enum$.is(value, "custom", "...")
Enum$.match(value, "custom", { ... })
```

Reply:
Resolved with https://github.com/peterboyer/unenum/pull/1/commits/0d49ab0aeba373875d9dfed30e0a0c06dc2a06e5.

The API is now, as per above:

```ts
Enum.builder(value, mapper?)
Enum.builder_(valuer, discriminant, mapper?)
Enum.is(value, matcher)
Enum.is_(value, discriminant, matcher)
Enum.match(value, matcher)
Enum.match_(value, matcher)
```